cmd/hujsonfmt/hujsonfmt.go

```diff
@@ -0,0 +1,189 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/hexops/gotextdiff"
+	"github.com/hexops/gotextdiff/myers"
+	"github.com/hexops/gotextdiff/span"
+	"github.com/tailscale/hujson"
+)
+
+var (
+	min   = flag.Bool("m", false, "minify results")
+	stand = flag.Bool("s", false, "standardize results to plain JSON")
+	diff  = flag.Bool("d", false, "display diffs instead of rewriting files")
+	list  = flag.Bool("l", false,
+		"list files whose formatting differs from hujsonfmt's",
+	)
+	write = flag.Bool("w", false,
+		"write result to (source) file instead of stdout",
+	)
+)
+
+func usage() {
+	fmt.Fprintf(os.Stderr, "usage: hujsonfmt [flags] [path ...]\n")
+	flag.PrintDefaults()
+}
+
+func main() {
+	err := mainE()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%s\n", err)
+		usage()
+		os.Exit(1)
+	}
+}
+
+func mainE() error {
+	flag.Usage = usage
+	flag.Parse()
+
+	ctx, cancel := trap(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	defer cancel()
+
+	files := flag.Args()
+
+	if len(files) == 0 || (len(files) == 1 && files[0] == "-") {
+		stat, _ := os.Stdin.Stat()
+		if (stat.Mode() & os.ModeCharDevice) != 0 {
+			return fmt.Errorf("no files paths or stdin provided")
+		}
+		if *write {
+			return fmt.Errorf("cannot use -w with standard input")
+		}
+
+		return processFile(ctx, "<standard input>", os.Stdin)
+	}
+
+	for _, file := range files {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		err := processFile(ctx, file, nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func processFile(ctx context.Context, filename string, in io.Reader) error {
+	src, err := readFile(filename, in)
+	if err != nil {
+		return err
+	}
+
+	if err = ctx.Err(); err != nil {
+		return err
+	}
+
+	// The main hujson functions will sometimes modify the original input byte
+	// slice. Hence we create a copy of the src byte slice to avoid modifying
+	// src, enabling us to reliably print diffs.
+	input := make([]byte, len(src))
+	_ = copy(input, src)
+
+	ouput, err := processSrc(input)
+	if err != nil {
+		return err
+	}
+
+	switch {
+	case *diff:
+		printDiff(filename, src, ouput)
+	case *list:
+		fmt.Println(filename)
+	case *write:
+		return os.WriteFile(filename, ouput, 0)
```

Comment:
Perhaps you want https://pkg.go.dev/tailscale.com/atomicfile

Reply:
I'm happy to use `atomicfile`, as long as the internal status of the package isn't an issue:

> This package should be considered internal; its API is not stable.

Reply:
I decided to use a backup from rather than the `atomicfile` package, as I figured various tools/editors might not handle the original file being wholesale replaced with `os.Rename()`, rather than just updating the exiting file in place.